README
[ $Id: README,v 1.14 2004/07/01 13:57:52 flovergine Exp $ ]

Yet Another Radius Daemon (YARD RADIUS) README File
---------------------------------------------------

This program is a RADIUS RFC-compliant daemon, which is derived from
original Livingston Enterprise Inc. (now Lucent Inc.) RADIUS daemon 
release 2.1. It adds a number of useful features to the LE daemon, i.e.

	Control of simultaneous logins.
	Support of Ascend, Cisco and USR boxes.
	Extended daily/monthly/yearly accouting information on a per-user basis.
	MD5 encrypted passwords support (both in passwd file and/or users file).
	Expirations in shadow file.
	Checking based on time-of-day, traffic and connection time.
	Support of PAM authentication and accounting.
	Binary form of accounting file.
	GDBM formats for users and user stats databases.
	Autoconfiguring capabilities of sources. 

It supports also all features of Lucent daemon, i.e.:

	Proxy RADIUS
	ActivCarda and iPass Support
	Accounting Signatures Now Required
	Vendor	Specific Attributes
	Virtual Ports
	Alternate Password File
	Address Binding
	Improved Messages
	Enhanced Debugging

All sources are much cleaner than the original versions, and
require an ANSI C compiler. A lots of potential buffer overflows have
been corrected by means of massive use of snprintf().

Directory content is:

BUGS			Some notes about possible bugs and how you can 
			submit bug reports.
README			This file.
README.LUCENT   	The original Lucent README file.
ChangeLog		Brief description of changes to the sources.
TODO			My personal to-do list for future releases.
CREDITS			File of credits.
INSTALL			Description of the installation procedure.
LICENSE			License of agreement file for YARD RADIUS.
LICENSE.LUCENT		Original License of agreement file from Lucent.
PGP-KEY			My public key generated with GnuPG.
configure		Autoconfiguring script

acconfig.h		All files used by GNU autoconf/automake
aclocal.m4
config.cache
config.guess
config.h.in
config.sub
configure
configure.in
shtool
install-sh
Makefile.am
Makefile.in
missing
mkinstalldirs
changeprefix
stamp-h.in
VERSION			Contains the release number

src/			Sources of YARD RADIUS programs.
doc/			Documentation of RADIUS protocol and YARD RADIUS (not
			complete).
pam/			Example of a PAM module for accounting.
conf/			Example of YARD RADIUS database and other stuff.
logs/			For installing.
include/		C Header files of the programs. 
man/			Man pages of all programs and extensions.

All software is under a BSD-like license. See LICENSE file for more
information. This is also the original license of Lucent Inc.

Please consult the INSTALL file and man pages to get
basic information about how you can build and use YARD RADIUS.

Home page: http://www.yardradius.org

YARD RADIUS project: https://sourceforge.net/projects/yardradius/

Mailing list: yardradius-users@lists.sourceforge.net

YARD Radius is also currently included in the current 
distribution of Debian GNU/Linux 3.0 (codename woody) or later version. 
I'm currently the maintainer of that stuff (and other pkgs) in the Debian Project.
The same deb package could also distributed on SF, but generally Debian files
are more up-to-date (for packaging, not functionalities). 


ChangeLog
[ $Id: ChangeLog,v 1.29 2004/07/01 13:57:52 flovergine Exp $ ]

(1.0.21) 2004-07-01 Francesco P. Lovergine <francesco@yardradius.org>

	The 'middle-land' release.

	  [Minor]
	 
	* Editorial changes in copyright notices of all source files.
	  Also some typos are now corrected.
	* New email and domain used extensively in every file.
	* Updated README to reflect new site URL and other info.
	* Refinement in size definitions of some data structures.

	  [Enhancement]
	  
	* Changed --version in -dumpversion for gcc versioning 
          in configure.in. This broke things in gcc 3. New arg is
          back compatible also.
	* Added entries limit to radlast. Thanks Russell Heilling 
	  for patch (see feature request 464791). Changed man page
	  as consequence.
	* Updated config.guess and config.sub, old releases could
	  break thing on current systems.

	  [Security]

	* A buffer overflow in calc_acctreq() has been patched.
	* process_menu() has been modified to check string limits.
	  Thanks Max Vozeler for pointing this so old issue and
	  showed a possible exploit.

(1.0.20) 2002-03-02 Francesco P. Lovergine <francesco@yardradius.org>
	* Documentation has been revised and updated. Many typos are
	  now corrected as well as some documentation incoherences.
	  New RFCs about the RADIUS protocol suite are enclosed and
	  obsolete ones removed.
	* Some files changed names for coherence.
	* getopt() returns int, not char.
	* SIGABRT is the standard definition in POSIX.1 which obsoletes
	  SIGIOT. In some environment the old signal is no more defined.
	* Now changeprefix script works good on Free BSD. The non
	  portable '-' argument has been skipped for sed. This caused
	  problem with configure script under that platform.
	* Dictionary updated with other VSAs (Ericsson and Xedia).
	* config.sub, config.guess and other automake scripts are now 
	  updated to the latest releases.
	* Major code revisions to avoid vulnerabilities due to strcpy()
          memcpy() and other dangerous functions. 
           - Digest Calculation Buffer Overflow Vulnerability.
           - Invalid attribute length calculation on malformed VSA.
          Non full-disclosures are available at CERT.org
	* Added my GPG key to code.

	  Some of those changes are stolen from a VA Linux patch for
	  Lucent Radius 2.1.

(1.0.19) 2001-08-27 Francesco P. Lovergine <francesco@yardradius.org>
	* Added a first version of man files under `man' directory.
	  Some old draft documentation files have been removed,
	  as consequence.
	* Default prefix is changed in `/usr/local/yardradius'
	* A couple of commas in configure.in which caused a ',6' 
	  file when --enable-debian was used has been deleted. 
	  A minor, but truly amazing bug :)
	* Many typos and grammar errors in doc corrected (or hoping so).
	* Configure.in did not install logs and doc directory when
	  prefix is not set on command line. Now also this is ok.

(1.0.18) 2001-08-21 Francesco P. Lovergine <francesco@yardradius.org>
	* Patched for a subdle bug due to __debug definition in extens.c.
	* Skipped auth_item check when Auth-Type=PAM.
	* Skipping void reply items in parsing GDBM users files.

	  All patches are due to William K. Volkman.

(1.0.17) 2001-08-18 Francesco P. Lovergine <francesco@yardradius.org> 
	* Configuration file modified to add a --enable-debian option.
	  It greatly simplifies support of Debian packaging. Debian
	  package is released as a separate package in Debian 'unstable'
	  distribution.
	* Now builddbm and radiusd are installend under sbin which
	  seems a more correct destination for those files.

(1.0.16/pre16) 2001-01-14 Francesco P. Lovergine <francesco@yardradius.org> 
	* Dictionary entries for BayNetworks/Nortel Annex is added.
	* Now, YARD RADIUS is compatible with GDBM 1.7.3, which seems the 
	  most commonly installed version. Just changed the gdbm_open()
	  function arguments for this.
	* Since now, the official name of the package is
	  Yard Radius. This to avoid confusion with other packages
	  with the same name, and to be coherent with a next-to-come 
	  Debian package (yeah!).
	  I changed also version numbering. The current version is 
	  1.0.16, i.e. 1.0 patch level 16. It seems more `standard',
	  and to be honest I really think this program is too much
	  robust to leave it in pre-release status... Am I an optimist?
	* Changed including of time.h and sys/time.h as suggested in autoconf
	  manuals. This allows installation under AIX and other systems
	  which could need both files. Incidentally, under AIX it compiles
	  also with the off-the-shelf C compiler.
	* PAM configuring in configure.in is changed to avoid a 
	  subdle configuration bug: some linux distributions do not
	  install PAM development libraries, so that PAM support cannot
	  be enabled on them, while building. This includes Debian 2.2
	  and RedHat.  That's not a problem on Solaris or FreeBSD 
	  which install them as default. This require a modification
	  of configure.in which is now smarter. 
	* An silly ortographic error in configure.in with --enable-pam
	  has been corrected.
	* Builddbm now works under the conf directory, not the current one.
	* License and Readme files have been changed. There is no more
	  an ANNOUNCE file and COPYING (which is the standard name for
	  GPL) is now named LICENSE (which is the standard name for BSD).
	* A CONFIGURATION text file has been added to summarize main
	  configuration issues. Are you happy?
	* Many other (little) adjustments.
	
(pre15) 2000-09-26 Francesco P. Lovergine <francesco@yardradius.org>
	* Installer now create also logs directory and explains
	  the structure.
	* Deleted Log entries from sources to avoid redundant comments
	  with CVS.

(pre14) 2000-09-21 Francesco P. Lovergine <francesco@yardradius.org>
	* Corrected autoconf/automake 
	  configuration to solve some problems of installation
	  under a few architectures. Configure now use Gnu Shell Tool

src/md5.c
/* MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm
 */

/* Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
 */

#include "yard.h"

/* Constants for MD5Transform routine.
 */
#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static void MD5Transform PROTO_LIST ((UINT4 [4], unsigned char [64]));
static void Encode PROTO_LIST
  ((unsigned char *, UINT4 *, unsigned int));
static void Decode PROTO_LIST
  ((UINT4 *, unsigned char *, unsigned int));
static void MD5_memcpy PROTO_LIST ((POINTER, POINTER, unsigned int));
static void MD5_memset PROTO_LIST ((POINTER, int, unsigned int));

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits.
 */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

void
md5_calc(output, input, inlen)
unsigned char *output;
unsigned char *input;                                /* input block */
unsigned int inlen;                     /* length of input block */
{
	MD5_CTX	context;

	MD5Init(&context);
	MD5Update(&context, input, inlen);
	MD5Final(output, &context);
}

/* MD5 initialization. Begins an MD5 operation, writing a new context.
 */
void MD5Init (context)
MD5_CTX *context;                                        /* context */
{
  context->count[0] = context->count[1] = 0;
  /* Load magic initialization constants.
*/
  context->state[0] = 0x67452301;
  context->state[1] = 0xefcdab89;
  context->state[2] = 0x98badcfe;
  context->state[3] = 0x10325476;
}

/* MD5 block update operation. Continues an MD5 message-digest
  operation, processing another message block, and updating the
  context.
 */
void MD5Update (context, input, inputLen)
MD5_CTX *context;                                        /* context */
unsigned char *input;                                /* input block */
unsigned int inputLen;                     /* length of input block */
{
  unsigned int i, index, partLen;

  /* Compute number of bytes mod 64 */
  index = (unsigned int)((context->count[0] >> 3) & 0x3F);

  /* Update number of bits */
  if ((context->count[0] += ((UINT4)inputLen << 3))
   < ((UINT4)inputLen << 3))
 context->count[1]++;
  context->count[1] += ((UINT4)inputLen >> 29);

  partLen = 64 - index;

  /* Transform as many times as possible.
*/
  if (inputLen >= partLen) {
 MD5_memcpy
   ((POINTER)&context->buffer[index], (POINTER)input, partLen);
 MD5Transform (context->state, context->buffer);

 for (i = partLen; i + 63 < inputLen; i += 64)
   MD5Transform (context->state, &input[i]);

 index = 0;
  }
  else
 i = 0;

  /* Buffer remaining input */
  MD5_memcpy
 ((POINTER)&context->buffer[index], (POINTER)&input[i],
  inputLen-i);
}

/* MD5 finalization. Ends an MD5 message-digest operation, writing the
  the message digest and zeroizing the context.
 */
void MD5Final (digest, context)
unsigned char digest[16];                         /* message digest */
MD5_CTX *context;                                       /* context */
{
  unsigned char bits[8];
  unsigned int index, padLen;

  /* Save number of bits */
  Encode (bits, context->count, 8);

  /* Pad out to 56 mod 64.
*/
  index = (unsigned int)((context->count[0] >> 3) & 0x3f);
  padLen = (index < 56) ? (56 - index) : (120 - index);
  MD5Update (context, PADDING, padLen);

  /* Append length (before padding) */
  MD5Update (context, bits, 8);

  /* Store state in digest */
  Encode (digest, context->state, 16);

  /* Zeroize sensitive information.
*/
  MD5_memset ((POINTER)context, 0, sizeof (*context));
}

/* MD5 basic transformation. Transforms state based on block.
 */
static void MD5Transform (state, block)
UINT4 state[4];
unsigned char block[64];
{
  UINT4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  Decode (x, block, 64);

  /* Round 1 */
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

 /* Round 2 */
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

  /* Round 4 */
  II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  /* Zeroize sensitive information.
*/
  MD5_memset ((POINTER)x, 0, sizeof (x));
}

/* Encodes input (UINT4) into output (unsigned char). Assumes len is
  a multiple of 4.
 */
static void Encode (output, input, len)
unsigned char *output;
UINT4 *input;
unsigned int len;
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
 output[j] = (unsigned char)(input[i] & 0xff);
 output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
 output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
 output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
  }
}

/* Decodes input (unsigned char) into output (UINT4). Assumes len is
  a multiple of 4.
 */
static void Decode (output, input, len)
UINT4 *output;
unsigned char *input;
unsigned int len;
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
 output[i] = ((UINT4)input[j]) | (((UINT4)input[j+1]) << 8) |
   (((UINT4)input[j+2]) << 16) | (((UINT4)input[j+3]) << 24);
}

/* Note: Replace "for loop" with standard memcpy if possible.
 */

static void MD5_memcpy (output, input, len)
POINTER output;
POINTER input;
unsigned int len;
{
  unsigned int i;

  for (i = 0; i < len; i++)
 output[i] = input[i];
}

/* Note: Replace "for loop" with standard memset if possible.
 */
static void MD5_memset (output, value, len)
POINTER output;
int value;
unsigned int len;
{
  unsigned int i;

  for (i = 0; i < len; i++)
 ((char *)output)[i] = (char)value;
}
